File: golang/hashmap.go
package hashmap

import (
	"errors"
)

const SIZE = 64

type Node struct {
	key  string
	val  string
	next *Node
}

type HashMap struct {
	Data []*Node
}

func NewHashMap() *HashMap {
	return &HashMap{Data: make([]*Node, SIZE)}
}

func Hash(key string) uint32 {
	total := uint32(0)
	for _, c := range key {
		total += uint32(c)
	}

	return total
}

func getIndex(key string) (int, error) {
	hash := Hash(key)

	index := hash % SIZE

	if index < 0 || index >= SIZE {
		return 0, errors.New("out of boundary")
	}

	return int(index), nil
}

func (h *HashMap) Put(key, value string) error {

	index, err := getIndex(key)
	if err != nil {
		return err
	}

	curr := h.Data[index]
	n := &Node{
		key:  key,
		val:  value,
		next: nil,
	}

	// if new data
	if curr == nil {
		h.Data[index] = n
		return nil
	}

	// if array not empty
	for {

		if curr.key == key {
			h.Data[index].val = value //replace
			return nil
		}

		if curr.next == nil {
			curr.next = n
			return nil
		}

		curr = curr.next
	}
}

func (h *HashMap) Get(key string) (string, error) {
	index, err := getIndex(key)
	if err != nil {
		return "", err
	}

	curr := h.Data[index]

	if curr == nil {
		return "", nil
	}

	for {
		if curr.key == key {
			return curr.val, nil
		}

		if curr.next == nil {
			return "", nil
		}
		curr = curr.next
	}
}

func (h *HashMap) Remove(key string) *string {
	index, err := getIndex(key)
	if err != nil {
		return nil
	}

	curr := h.Data[index]

	if curr == nil {
		return nil
	}

	prev := curr

	for {
		if curr.key == key {
			if prev == curr && curr != nil && curr.next != nil {
				h.Data[index] = curr.next
				return &curr.val
			}

			if prev == curr && curr != nil && curr.next == nil {
				h.Data[index] = nil
				return &curr.val
			}

			if prev != curr && curr.next != nil {
				prev.next = curr.next
				return &curr.val
			}

			if prev != curr && curr.next == nil {
				prev.next = nil
				return &curr.val
			}
		}

		if curr.next == nil {
			return nil
		}

		prev = curr
		curr = curr.next
	}
}

func (h *HashMap) IsEmpty() bool {
	for _, v := range h.Data {
		if v != nil {
			return false
		}
	}

	return true
}

File: java/com/bondhan/java/HashMap.java
package com.bondhan.java;

import java.io.*;

public class HashMap {

    public final int SIZE = 64;

    private Node[] Nodes;

    public HashMap() {
        Nodes = new Node[SIZE];
    }

    public HashMap(int initial) {
        Nodes = new Node[initial];
    }

    public int hash(String key) {
        int total = 0;
        for (int i = 0; i < key.length(); i++) {
            total += key.codePointAt(i);
        }

        return total;
    }

    public int getIndex(String key) {
        int h = hash(key);

        return h & (SIZE - 1);
    }

    public String Put(String key, String value) {
        int index = getIndex(key);

        Node n = new Node(key, value);

        if (Nodes[index] == null) {
            Nodes[index] = n;
            return value;
        }

        Node curr = Nodes[index];

        while (curr != null) {
            if (Objects.equals(curr.Key, key)) {
                curr.Value = value;
                return value;
            }

            if (curr.Next == null) {
                curr.Next = n;
                return value;
            }

            curr = curr.Next;
        }

        return value;
    }

    public String Get(String key) {
        int index = getIndex(key);

        Node curr = Nodes[index];

        while (curr != null) {

            if (Objects.equals(curr.Key, key)) {
                return curr.Value;
            }

            if (curr.Next == null) {
                return "";
            }

            curr = curr.Next;
        }

        return "";
    }

    public String Remove(String key) {
        int index = getIndex(key);

        Node curr = Nodes[index];
        Node prev = curr;

        while (curr != null) {

            if (Objects.equals(curr.Key, key)){
                if (prev == curr) {
                    String val = Nodes[index].Value;
                    Nodes[index] = null;
                    return val;
                }

                if (prev != curr) {
                    if (curr.Next != null) {
                        prev.Next = curr.Next;
                        return curr.Value;
                    }

                    prev.Next = null;
                    return curr.Value;
                }
            }

            if (curr.Next != null) {
                prev = curr;
                curr = curr.Next;
            }
        }

        return "";
    }
}